OFX export fails with FileNotFoundException

Exporting transactions in OFX format fail if there's a difference in the wall clock time between the start and end of the export.

```
E/ExportAsyncTask: /data/data/org.gnucash.android.devel/cache/OFX/20151204_175635_gnucash_export.ofx (No such file or directory)
E/SchedulerService: org.gnucash.android.export.Exporter$ExporterException: Failed to generate Open Financial eXchange-/data/data/org.gnucash.android.devel/cache/OFX/20151204_175635_gnucash_export.ofx (No such file or directory)
W/dalvikvm: threadid=17: thread exiting with uncaught exception (group=0x401e1560)
E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
                  java.lang.RuntimeException: An error occured while executing doInBackground()
                      at android.os.AsyncTask$3.done(AsyncTask.java:200)
                      at java.util.concurrent.FutureTask$Sync.innerSetException(FutureTask.java:274)
                      at java.util.concurrent.FutureTask.setException(FutureTask.java:125)
                      at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:308)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:138)
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1088)
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:581)
                      at java.lang.Thread.run(Thread.java:1019)
                   Caused by: org.gnucash.android.export.Exporter$ExporterException: Failed to generate Open Financial eXchange-/data/data/org.gnucash.android.devel/cache/OFX/20151204_175635_gnucash_export.ofx (No such file or directory)
                      at org.gnucash.android.export.ExportAsyncTask.moveExportToSDCard(ExportAsyncTask.java:355)
                      at org.gnucash.android.export.ExportAsyncTask.doInBackground(ExportAsyncTask.java:184)
                      at org.gnucash.android.export.ExportAsyncTask.doInBackground(ExportAsyncTask.java:79)
                      at android.os.AsyncTask$2.call(AsyncTask.java:185)
                      at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:306)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:138) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1088) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:581) 
                      at java.lang.Thread.run(Thread.java:1019) 
                   Caused by: java.io.FileNotFoundException: /data/data/org.gnucash.android.devel/cache/OFX/20151204_175635_gnucash_export.ofx (No such file or directory)
                      at org.apache.harmony.luni.platform.OSFileSystem.open(Native Method)
                      at dalvik.system.BlockGuard$WrappedFileSystem.open(BlockGuard.java:232)
                      at java.io.FileInputStream.<init>(FileInputStream.java:80)
                      at org.gnucash.android.export.ExportAsyncTask.moveFile(ExportAsyncTask.java:457)
                      at org.gnucash.android.export.ExportAsyncTask.moveExportToSDCard(ExportAsyncTask.java:350)
                      at org.gnucash.android.export.ExportAsyncTask.doInBackground(ExportAsyncTask.java:184) 
                      at org.gnucash.android.export.ExportAsyncTask.doInBackground(ExportAsyncTask.java:79) 
                      at android.os.AsyncTask$2.call(AsyncTask.java:185) 
                      at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:306) 
                      at java.util.concurrent.FutureTask.run(FutureTask.java:138) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1088) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:581) 
                      at java.lang.Thread.run(Thread.java:1019) 
```

It happens when the second call to Export.getExportCacheFilePath() in
OfxExporter.generateExport() returns a file name with a time stamp different to
the first call in the same method. The root of the problem is
getExportCacheFilePath(), which builds the file name with the current time.

Received through Crashlitics as issue `#129`.

--- a/app/src/main/java/org/gnucash/android/export/Exporter.java
+++ b/app/src/main/java/org/gnucash/android/export/Exporter.java
@@ -58,7 +58,7 @@ public abstract class Exporter {
     /**
      * Application folder on external storage
      */
-    public static final String BASE_FOLDER_PATH = Environment.getExternalStorageDirectory() + "/" + BuildConfig.APPLICATION_ID;
+    private static final String BASE_FOLDER_PATH = Environment.getExternalStorageDirectory() + "/" + BuildConfig.APPLICATION_ID;
 
     /**
      * Folder where exports like QIF and OFX will be saved for access by external programs
@@ -73,7 +73,7 @@ public abstract class Exporter {
     /**
      * Export options
      */
-    protected ExportParams mExportParams;
+    protected final ExportParams mExportParams;
 
     /**
      * Cache directory to which files will be first exported before moved to final destination.
@@ -82,7 +82,7 @@ public abstract class Exporter {
      *    The files created here are only accessible within this application, and should be copied to SD card before they can be shared
      * </p>
      */
-    protected File mCacheDir;
+    private final File mCacheDir;
 
     private static final SimpleDateFormat EXPORT_FILENAME_DATE_FORMAT = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US);
 
@@ -96,13 +96,13 @@ public abstract class Exporter {
      * Adapter for retrieving accounts to export
      * Subclasses should close this object when they are done with exporting
      */
-    protected AccountsDbAdapter mAccountsDbAdapter;
-    protected TransactionsDbAdapter mTransactionsDbAdapter;
-    protected SplitsDbAdapter mSplitsDbAdapter;
-    protected ScheduledActionDbAdapter mScheduledActionDbAdapter;
-    protected PricesDbAdapter mPricesDbAdapter;
-    protected CommoditiesDbAdapter mCommoditiesDbAdapter;
-    protected Context mContext;
+    protected final AccountsDbAdapter mAccountsDbAdapter;
+    protected final TransactionsDbAdapter mTransactionsDbAdapter;
+    protected final SplitsDbAdapter mSplitsDbAdapter;
+    protected final ScheduledActionDbAdapter mScheduledActionDbAdapter;
+    protected final PricesDbAdapter mPricesDbAdapter;
+    protected final CommoditiesDbAdapter mCommoditiesDbAdapter;
+    protected final Context mContext;
     private String mExportCacheFilePath;
 
     public Exporter(ExportParams params, SQLiteDatabase db) {

--- a/app/src/main/java/org/gnucash/android/export/Exporter.java
+++ b/app/src/main/java/org/gnucash/android/export/Exporter.java
@@ -103,6 +103,7 @@ public abstract class Exporter {
     protected PricesDbAdapter mPricesDbAdapter;
     protected CommoditiesDbAdapter mCommoditiesDbAdapter;
     protected Context mContext;
+    private String mExportCacheFilePath;
 
     public Exporter(ExportParams params, SQLiteDatabase db) {
         this.mExportParams = params;
@@ -123,6 +124,7 @@ public abstract class Exporter {
             mCommoditiesDbAdapter = new CommoditiesDbAdapter(db);
         }
 
+        mExportCacheFilePath = null;
         mCacheDir = new File(mContext.getCacheDir(), params.getExportFormat().name());
         mCacheDir.mkdir();
         purgeDirectory(mCacheDir);
@@ -184,10 +186,16 @@ public abstract class Exporter {
      * @return Absolute path to file
      */
     protected String getExportCacheFilePath(){
-        String cachePath = mCacheDir.getAbsolutePath();
-        if (!cachePath.endsWith("/"))
-            cachePath += "/";
-        return cachePath + buildExportFilename(mExportParams.getExportFormat());
+        // The file name contains a timestamp, so ensure it doesn't change with multiple calls to
+        // avoid issues like #448
+        if (mExportCacheFilePath == null) {
+            String cachePath = mCacheDir.getAbsolutePath();
+            if (!cachePath.endsWith("/"))
+                cachePath += "/";
+            mExportCacheFilePath = cachePath + buildExportFilename(mExportParams.getExportFormat());
+        }
+
+        return mExportCacheFilePath;
     }
 
     /**

